From 43ac1c4ba1ec075bceb7ed3c34ecfb673e2d3dbf Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Thu, 24 Apr 2025 15:56:11 +0200
Subject: [PATCH 825/841] feat(st): add check to function linked to NVMEM
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Ensure error codes from NVMEM framework functions are propagated and
checked.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 4c9f7574a4eac6c3004f50029727df55596d793d feat(st): add check to function linked to NVMEM]
Change-Id: I081c6b9bebe558b4906743f4b6897e290861469d
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/453851
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/common/bl2_io_storage.c         |  35 ++++---
 plat/st/common/include/stm32mp_common.h |  12 +--
 plat/st/common/stm32mp_common.c         | 127 ++++++++++++++++++------
 plat/st/stm32mp1/bl2_plat_setup.c       |   9 +-
 plat/st/stm32mp2/bl2_plat_setup.c       |   9 +-
 5 files changed, 136 insertions(+), 56 deletions(-)

diff --git a/plat/st/common/bl2_io_storage.c b/plat/st/common/bl2_io_storage.c
index 1a6395d5f..9a02edfe9 100644
--- a/plat/st/common/bl2_io_storage.c
+++ b/plat/st/common/bl2_io_storage.c
@@ -794,6 +794,7 @@ uint32_t plat_fwu_get_boot_idx(void)
 	 * even if this function is called several times.
 	 */
 	static uint32_t boot_idx = INVALID_BOOT_IDX;
+	int err = 0;
 
 	if (boot_idx == INVALID_BOOT_IDX) {
 		const struct fwu_metadata *data = fwu_get_metadata();
@@ -803,20 +804,22 @@ uint32_t plat_fwu_get_boot_idx(void)
 
 		switch (data->bank_state[boot_idx]) {
 		case FWU_BANK_STATE_ACCEPTED:
-			stm32_set_max_fwu_trial_boot_cnt();
+			err = stm32_set_max_fwu_trial_boot_cnt();
 			break;
 		case FWU_BANK_STATE_VALID:
-			bootcount = stm32_get_and_dec_fwu_trial_boot_cnt();
-			if (bootcount == 1U) {
-				WARN("Trial FWU fails %u times\n",
-				     (FWU_MAX_TRIAL_REBOOT - 1U));
-				boot_idx = fwu_get_alternate_boot_bank();
-			} else if (bootcount == 0U) {
-				WARN("Trial backup register empty : set max boot count\n");
-				stm32_set_max_fwu_trial_boot_cnt();
-			} else {
-				VERBOSE("Trial FWU: %u\n",
-					FWU_MAX_TRIAL_REBOOT - bootcount);
+			err = stm32_get_and_dec_fwu_trial_boot_cnt(&bootcount);
+			if (err == 0) {
+				if (bootcount == 1U) {
+					WARN("Trial FWU fails %u times\n",
+					     (FWU_MAX_TRIAL_REBOOT - 1U));
+					boot_idx = fwu_get_alternate_boot_bank();
+				} else if (bootcount == 0U) {
+					WARN("Trial backup register empty : set max boot count\n");
+					err = stm32_set_max_fwu_trial_boot_cnt();
+				} else {
+					VERBOSE("Trial FWU: %u\n",
+						FWU_MAX_TRIAL_REBOOT - bootcount);
+				}
 			}
 			break;
 		case FWU_BANK_STATE_INVALID:
@@ -824,9 +827,15 @@ uint32_t plat_fwu_get_boot_idx(void)
 			ERROR("The active bank(%u) of the platform is in Invalid State.\n",
 			      boot_idx);
 			boot_idx = fwu_get_alternate_boot_bank();
-			stm32_clear_fwu_trial_boot_cnt();
+			err = stm32_clear_fwu_trial_boot_cnt();
 			break;
 		}
+
+		if (err != 0) {
+			ERROR("%s: Bkp register access failed. Bank state: %d\n",
+				__func__, data->bank_state[boot_idx]);
+			panic();
+		}
 	}
 
 	return boot_idx;
diff --git a/plat/st/common/include/stm32mp_common.h b/plat/st/common/include/stm32mp_common.h
index 567ca0481..39660a963 100644
--- a/plat/st/common/include/stm32mp_common.h
+++ b/plat/st/common/include/stm32mp_common.h
@@ -166,9 +166,9 @@ int stm32mp_map_retram(void);
 int stm32mp_unmap_retram(void);
 
 /* Function to save boot info */
-void stm32_save_boot_info(boot_api_context_t *boot_context);
+int stm32_save_boot_info(boot_api_context_t *boot_context);
 /* Function to get boot peripheral info */
-void stm32_get_boot_interface(uint32_t *interface, uint32_t *instance);
+int stm32_get_boot_interface(uint32_t *interface, uint32_t *instance);
 
 /* Display board information from the value found in OTP fuse */
 void stm32_display_board_info(uint32_t board_id);
@@ -189,10 +189,10 @@ int stm32_get_stop2_entrypoint_cell(struct nvmem_cell *stop2_entrypoint);
 #endif
 
 #if PSA_FWU_SUPPORT
-void stm32_fwu_set_boot_idx(void);
-uint32_t stm32_get_and_dec_fwu_trial_boot_cnt(void);
-void stm32_set_max_fwu_trial_boot_cnt(void);
-void stm32_clear_fwu_trial_boot_cnt(void);
+int stm32_fwu_set_boot_idx(void);
+int stm32_get_and_dec_fwu_trial_boot_cnt(uint32_t *cnt);
+int stm32_set_max_fwu_trial_boot_cnt(void);
+int stm32_clear_fwu_trial_boot_cnt(void);
 #endif /* PSA_FWU_SUPPORT */
 
 #endif /* STM32MP_COMMON_H */
diff --git a/plat/st/common/stm32mp_common.c b/plat/st/common/stm32mp_common.c
index c21bd8cde..b4d780559 100644
--- a/plat/st/common/stm32mp_common.c
+++ b/plat/st/common/stm32mp_common.c
@@ -351,6 +351,7 @@ int stm32mp_uart_console_setup(void)
 	struct dt_node_info dt_uart_info;
 	uint32_t clk_rate = 0U;
 	int result;
+	int ret __maybe_unused;
 	uint32_t boot_itf __unused;
 	uint32_t boot_instance __unused;
 
@@ -369,10 +370,9 @@ int stm32mp_uart_console_setup(void)
 	console_clk = dt_uart_info.clock;
 
 #if (STM32MP_UART_PROGRAMMER || !defined(IMAGE_BL2)) && !STM32MP_SSP
-	stm32_get_boot_interface(&boot_itf, &boot_instance);
-
-	if ((boot_itf == BOOT_API_CTX_BOOT_INTERFACE_SEL_SERIAL_UART) &&
-	    (get_uart_address(boot_instance) == dt_uart_info.base)) {
+	if ((stm32_get_boot_interface(&boot_itf, &boot_instance) != 0) ||
+	    ((boot_itf == BOOT_API_CTX_BOOT_INTERFACE_SEL_SERIAL_UART) &&
+	    (get_uart_address(boot_instance) == dt_uart_info.base))) {
 		return -EACCES;
 	}
 #endif
@@ -449,7 +449,7 @@ void stm32_display_board_info(uint32_t board_id)
 }
 
 #if !STM32MP_SSP
-void stm32_save_boot_info(boot_api_context_t *boot_context)
+int stm32_save_boot_info(boot_api_context_t *boot_context)
 {
 	uint32_t auth_status;
 	struct nvmem_cell boot_mode = {};
@@ -457,6 +457,7 @@ void stm32_save_boot_info(boot_api_context_t *boot_context)
 	uint32_t clear = BOOT_ITF_MASK | BOOT_INST_MASK | BOOT_PART_MASK |
 			 BOOT_AUTH_MASK;
 	uint32_t set;
+	int ret;
 
 	assert(boot_context->boot_interface_instance <= (BOOT_INST_MASK >> BOOT_INST_SHIFT));
 	assert(boot_context->boot_interface_selected <= (BOOT_ITF_MASK >> BOOT_ITF_SHIFT));
@@ -482,28 +483,42 @@ void stm32_save_boot_info(boot_api_context_t *boot_context)
 	      (boot_context->boot_partition_used_toboot << BOOT_PART_SHIFT) |
 	      (auth_status << BOOT_AUTH_SHIFT);
 
-	stm32_get_boot_mode_cell(&boot_mode);
+	ret = stm32_get_boot_mode_cell(&boot_mode);
+	if (ret != 0) {
+		return ret;
+	}
 	nvmem_cell_read(&boot_mode, (uint8_t *)&reg_val, sizeof(reg_val), NULL);
 	reg_val &= ~clear;
 	reg_val |= set;
 	nvmem_cell_write(&boot_mode, (uint8_t *)&reg_val, sizeof(reg_val));
+
+	return 0;
 }
 
-void stm32_get_boot_interface(uint32_t *interface, uint32_t *instance)
+int stm32_get_boot_interface(uint32_t *interface, uint32_t *instance)
 {
 	static uint32_t itf;
 	struct nvmem_cell boot_mode = {};
 	uint32_t reg_val = 0;
 
 	if (itf == 0U) {
-		stm32_get_boot_mode_cell(&boot_mode);
-		nvmem_cell_read(&boot_mode, (uint8_t *)&reg_val,
-				sizeof(reg_val), NULL);
+		int ret = stm32_get_boot_mode_cell(&boot_mode);
+
+		if (ret != 0) {
+			return ret;
+		}
+		ret = nvmem_cell_read(&boot_mode, (uint8_t *)&reg_val,
+				      sizeof(reg_val), NULL);
+		if (ret != 0) {
+			return ret;
+		}
 		itf = reg_val & (BOOT_ITF_MASK | BOOT_INST_MASK);
 	}
 
 	*interface = (itf & BOOT_ITF_MASK) >> BOOT_ITF_SHIFT;
 	*instance = (itf & BOOT_INST_MASK) >> BOOT_INST_SHIFT;
+
+	return 0;
 }
 
 static int stm32_get_bootinfo_cell(const char *name, struct nvmem_cell *cell)
@@ -533,7 +548,11 @@ int stm32_get_magic_number_cell(struct nvmem_cell *magic_number)
 	static struct nvmem_cell s_magic_number = { 0 };
 
 	if (!initialized) {
-		stm32_get_bootinfo_cell("magic-number", &s_magic_number);
+		int ret = stm32_get_bootinfo_cell("magic-number", &s_magic_number);
+
+		if (ret != 0) {
+			return ret;
+		}
 		initialized = true;
 	}
 
@@ -548,7 +567,11 @@ int stm32_get_core1_branch_address_cell(struct nvmem_cell *core1_branch_address)
 	static struct nvmem_cell s_core1_branch_address = { 0 };
 
 	if (!initialized) {
-		stm32_get_bootinfo_cell("core1-branch-address", &s_core1_branch_address);
+		int ret = stm32_get_bootinfo_cell("core1-branch-address", &s_core1_branch_address);
+
+		if (ret != 0) {
+			return ret;
+		}
 		initialized = true;
 	}
 
@@ -564,7 +587,11 @@ int stm32_get_fwu_info_cell(struct nvmem_cell *fwu_info)
 	static struct nvmem_cell s_fwu_info = { 0 };
 
 	if (!initialized) {
-		stm32_get_bootinfo_cell("fwu-info", &s_fwu_info);
+		int ret = stm32_get_bootinfo_cell("fwu-info", &s_fwu_info);
+
+		if (ret != 0) {
+			return ret;
+		}
 		initialized = true;
 	}
 
@@ -579,7 +606,11 @@ int stm32_get_boot_mode_cell(struct nvmem_cell *boot_mode)
 	static struct nvmem_cell s_boot_mode = { 0 };
 
 	if (!initialized) {
-		stm32_get_bootinfo_cell("boot-mode", &s_boot_mode);
+		int ret = stm32_get_bootinfo_cell("boot-mode", &s_boot_mode);
+
+		if (ret != 0) {
+			return ret;
+		}
 		initialized = true;
 	}
 
@@ -595,7 +626,11 @@ int stm32_get_stop2_entrypoint_cell(struct nvmem_cell *stop2_entrypoint)
 	static struct nvmem_cell s_stop2_entrypoint = { 0 };
 
 	if (!initialized) {
-		stm32_get_bootinfo_cell("stop2-entrypoint", &s_stop2_entrypoint);
+		int ret = stm32_get_bootinfo_cell("stop2-entrypoint", &s_stop2_entrypoint);
+
+		if (ret != 0) {
+			return ret;
+		}
 		initialized = true;
 	}
 
@@ -627,58 +662,84 @@ static int stm32_nvmem_cell_clrset(struct nvmem_cell *cell, uint32_t clear,
 	return 0;
 }
 
-void stm32_fwu_set_boot_idx(void)
+int stm32_fwu_set_boot_idx(void)
 {
 	struct nvmem_cell fwu_info = {};
+	int ret = 0;
 
 	uint32_t clear = FWU_INFO_IDX_MSK;
 	uint32_t set = (plat_fwu_get_boot_idx() << FWU_INFO_IDX_OFF) &
 		       FWU_INFO_IDX_MSK;
 
-	stm32_get_fwu_info_cell(&fwu_info);
+	ret = stm32_get_fwu_info_cell(&fwu_info);
+	if (ret != 0) {
+		return ret;
+	}
 
-	stm32_nvmem_cell_clrset(&fwu_info, clear, set);
+	return stm32_nvmem_cell_clrset(&fwu_info, clear, set);
 }
 
-uint32_t stm32_get_and_dec_fwu_trial_boot_cnt(void)
+int stm32_get_and_dec_fwu_trial_boot_cnt(uint32_t *cnt)
 {
 	struct nvmem_cell fwu_info_cell = {};
 	uint32_t try_cnt;
 	uint32_t fwu_info = 0;
 
-	stm32_get_fwu_info_cell(&fwu_info_cell);
+	int ret = stm32_get_fwu_info_cell(&fwu_info_cell);
+
+	if (ret != 0) {
+		return ret;
+	}
 
-	nvmem_cell_read(&fwu_info_cell, (uint8_t *)&fwu_info, sizeof(fwu_info),
-			NULL);
+	ret = nvmem_cell_read(&fwu_info_cell, (uint8_t *)&fwu_info,
+			      sizeof(fwu_info), NULL);
+	if (ret != 0) {
+		return ret;
+	}
 
 	try_cnt = (fwu_info & FWU_INFO_CNT_MSK) >> FWU_INFO_CNT_OFF;
 
 	assert(try_cnt <= FWU_MAX_TRIAL_REBOOT);
 
 	if (try_cnt != 0U) {
-		stm32_nvmem_cell_clrset(&fwu_info_cell, FWU_INFO_CNT_MSK,
-					(try_cnt - 1U) << FWU_INFO_CNT_OFF);
+		ret = stm32_nvmem_cell_clrset(&fwu_info_cell, FWU_INFO_CNT_MSK,
+					      (try_cnt - 1U) << FWU_INFO_CNT_OFF);
+		if (ret != 0) {
+			return ret;
+		}
 	}
 
-	return try_cnt;
+	*cnt = try_cnt;
+
+	return 0;
 }
 
-void stm32_set_max_fwu_trial_boot_cnt(void)
+int stm32_set_max_fwu_trial_boot_cnt(void)
 {
 	struct nvmem_cell fwu_info_cell = {};
 
-	stm32_get_fwu_info_cell(&fwu_info_cell);
-	stm32_nvmem_cell_clrset(&fwu_info_cell, FWU_INFO_CNT_MSK,
-				(FWU_MAX_TRIAL_REBOOT << FWU_INFO_CNT_OFF) &
-					FWU_INFO_CNT_MSK);
+	int ret = stm32_get_fwu_info_cell(&fwu_info_cell);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	return stm32_nvmem_cell_clrset(&fwu_info_cell, FWU_INFO_CNT_MSK,
+				       (FWU_MAX_TRIAL_REBOOT << FWU_INFO_CNT_OFF) &
+				       FWU_INFO_CNT_MSK);
 }
 
-void stm32_clear_fwu_trial_boot_cnt(void)
+int stm32_clear_fwu_trial_boot_cnt(void)
 {
 	struct nvmem_cell fwu_info_cell = {};
 
-	stm32_get_fwu_info_cell(&fwu_info_cell);
-	stm32_nvmem_cell_clrset(&fwu_info_cell, FWU_INFO_CNT_MSK, 0U);
+	int ret = stm32_get_fwu_info_cell(&fwu_info_cell);
+
+	if (ret != 0) {
+		return ret;
+	}
+
+	return stm32_nvmem_cell_clrset(&fwu_info_cell, FWU_INFO_CNT_MSK, 0U);
 }
 #endif /* PSA_FWU_SUPPORT */
 #endif /* !STM32MP_SSP */
diff --git a/plat/st/stm32mp1/bl2_plat_setup.c b/plat/st/stm32mp1/bl2_plat_setup.c
index 05e525dc5..83013c902 100644
--- a/plat/st/stm32mp1/bl2_plat_setup.c
+++ b/plat/st/stm32mp1/bl2_plat_setup.c
@@ -409,7 +409,9 @@ void bl2_el3_plat_arch_setup(void)
 		panic();
 	}
 
-	stm32_save_boot_info(boot_context);
+	if (stm32_save_boot_info(boot_context) != 0) {
+		panic();
+	}
 
 #if STM32MP_USB_PROGRAMMER && STM32MP15
 	/* Deconfigure all UART RX pins configured by ROM code */
@@ -735,7 +737,10 @@ int bl2_plat_handle_post_image_load(unsigned int image_id)
 		bl32_mem_params->ep_info.lr_svc = bl_mem_params->ep_info.pc;
 #if PSA_FWU_SUPPORT
 		if (plat_fwu_is_enabled()) {
-			stm32_fwu_set_boot_idx();
+			err = stm32_fwu_set_boot_idx();
+			if (err != 0) {
+				panic();
+			}
 		}
 #endif /* PSA_FWU_SUPPORT */
 		break;
diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 0abda2f50..bbf709de4 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -498,7 +498,9 @@ void bl2_el3_plat_arch_setup(void)
 	}
 #endif
 
-	stm32_save_boot_info(boot_context);
+	if (stm32_save_boot_info(boot_context) != 0) {
+		panic();
+	}
 
 	/* Masking potential tamper during BL2 */
 	stm32mp_syscfg_mask_potential_tamper_enable();
@@ -808,7 +810,10 @@ int bl2_plat_handle_post_image_load(unsigned int image_id)
 
 	case BL33_IMAGE_ID:
 #if PSA_FWU_SUPPORT
-		stm32_fwu_set_boot_idx();
+		err = stm32_fwu_set_boot_idx();
+		if (err != 0) {
+			panic();
+		}
 #endif /* PSA_FWU_SUPPORT */
 		break;
 
-- 
2.47.2

